pip/pip-297.md

```diff
@@ -0,0 +1,168 @@
+  incompatible system. The function framework will catch these exceptions, report them to users, and terminate the
+  function.
+
+# Motivation
+
+The current implementation of the connector and Pulsar Function exception handler cannot handle the unrecoverable
+exceptions that are thrown outside of the function instance thread.
+
+For example, suppose we have a sink connector that uses its own threads to batch-sink the data to an external system. If
+any unrecoverable exceptions occur in those threads, the function instance thread will not be aware of them and will
+not be able to terminate the connector. This will cause the connector to hang indefinitely. There is a related issue
+here: https://github.com/apache/pulsar/issues/9464
+
+The same problem exists for the source connector. The source connector may also use a separate thread to fetch data from
+an external system. If any fatal exceptions happen in that thread, the connector will also hang forever. This issue has
+been observed for the Kafka source connector: https://github.com/apache/pulsar/issues/9464. We have fixed it by adding
+the notifyError method to the PushSource class in PIP-281: https://github.com/apache/pulsar/pull/20807. However, this
+does not solve the same problem that all source connectors face.
+
+The problem is same for the Pulsar Function. The user can also process messages asynchronously in a Pulsar Function. If
+any fatal exceptions happen in the user's thread, the function framework can't handle the exception
+correctly.
+
+We need a way for the connector and function developers to throw unrecoverable exceptions outside the function instance
+thread. The function framework should catch these exceptions and terminate the function accordingly.
+
+# Goals
+
+## In Scope
+
+- This proposal will apply both to the Pulsar Function and the Pulsar Connector.
+- Support throwing unrecoverable exceptions outside the function instance thread.
+- The function framework should terminate the connector or function when there are any unrecoverable exceptions thrown
+  asynchronously.
+
+## Out of Scope
+
+- The fixes of the exception-raising issue mentioned in the Motivation part for all the connectors are not included in
+  this PIP. This PIP only provides the feature for the connector developer to raise the exception outside the function
+  instance thread. The fixes should be in serval different PRs.
+
+# High Level Design
+
+Introduce a new method `raiseException` to the context. All the connector implementation code and the function code can
+use this context and call the `raiseException` method to raise an unrecoverable exception to the function framework.
+
+After the connector or function raises the unrecoverable exception, the function instance thread will be interrupted.
+The function framework then could catch the exception, log it, and then terminate the function instance.
```

Comment:
Is it possible for users to terminate the function instance directly for now?

Reply:
No. There is no direct way to terminate it. PIP-281 uses an indirect way to implement it by resending an exception record to the source connector.